packages/apollo-codegen-typescript/src/codeGeneration.ts

```diff
@@ -246,6 +335,68 @@ export class TypescriptAPIGenerator extends TypescriptGenerator {
     this.scopeStackPop();
   }
 
+  public getGlobalTypesUsedForOperation(doc: Operation | Fragment, context: CompilerContext) {
+    let docTypesUsed: GraphQLType[] = [];
+
+    if (doc.hasOwnProperty('operationName')) {
+      const operation = doc as Operation;
+      docTypesUsed = operation.variables.map(({ type }) => type);
+    }
+
+    const reduceTypesForDocument = (
+      nested: SelectionSet,
+      acc: GraphQLType[]
+    ) => {
+      acc = nested.selections
+        .reduce((selectionAcc, selection) => {
+          switch (selection.kind) {
+            case 'Field':
+            case 'TypeCondition':
+              if (selection.selectionSet) {
+                selectionAcc = reduceTypesForDocument(selection.selectionSet, selectionAcc);
+              } else {
+                selectionAcc = maybePush(selectionAcc, selection.type);
+              }
+              break;
+            case 'FragmentSpread':
+              selectionAcc = reduceTypesForDocument(selection.selectionSet, selectionAcc);
+              break;
+            default:
+              break;
+          }
+
+          return selectionAcc;
+        }, acc);
+
+      return acc;
+    }
+
+    docTypesUsed = reduceTypesForDocument(doc.selectionSet, docTypesUsed)
+      .reduce(this.reduceGlobalTypesUsed, []);
+
+    return context.typesUsed.filter((type) => {
+      return (type instanceof GraphQLEnumType || type instanceof GraphQLInputObjectType)
+        && docTypesUsed.find((typeUsed) => type === typeUsed);
+    });
+  }
+
+  private reduceGlobalTypesUsed = (
+    acc: (GraphQLType | GraphQLOutputType)[],
+    type: GraphQLType
+  ) =>{
+    if (type instanceof GraphQLNonNull) {
+      type = getNullableType(type);
+    }
+
+    if (type instanceof GraphQLList) {
+      type = type.ofType
+    }
+
+    acc = maybePush(acc, type);
```

Comment:
Same as above, could this be replaced by an immutable update `[...acc, type]`?

Reply:
As far as I can tell, the intention of `maybePush` from the original `reduceTypesUsed` was that there are no duplicated types in `acc` as `maybePush` checks for dups. However since we filter later anyway, we could accumulate duplicates here. Or we could collect them as an object instead thus having O(1) later for the lookups.